src/network/network_crypto.cpp

```diff
@@ -0,0 +1,401 @@
+ */
+bool X25519DerivedKeys::Exchange(const X25519PublicKey &peer_public_key, X25519KeyExchangeSide side,
+		const X25519SecretKey &our_secret_key, const X25519PublicKey &our_public_key, std::string_view extra_payload)
+{
+	X25519Key shared_secret;
+	crypto_x25519(shared_secret.data(), our_secret_key.data(), peer_public_key.data());
+	if (std::all_of(shared_secret.begin(), shared_secret.end(), [](auto v) { return v == 0; })) {
+		/* A shared secret of all zeros means that the peer tried to force the shared secret to a known constant. */
+		return false;
+	}
+
+	crypto_blake2b_ctx ctx;
+	crypto_blake2b_init(&ctx, this->keys.size());
+	crypto_blake2b_update(&ctx, shared_secret.data(), shared_secret.size());
+	switch (side) {
+		case X25519KeyExchangeSide::SERVER:
+			crypto_blake2b_update(&ctx, our_public_key.data(), our_public_key.size());
+			crypto_blake2b_update(&ctx, peer_public_key.data(), peer_public_key.size());
+			break;
+		case X25519KeyExchangeSide::CLIENT:
+			crypto_blake2b_update(&ctx, peer_public_key.data(), peer_public_key.size());
+			crypto_blake2b_update(&ctx, our_public_key.data(), our_public_key.size());
+			break;
+		default:
+			NOT_REACHED();
+	}
+	crypto_blake2b_update(&ctx, reinterpret_cast<const uint8_t *>(extra_payload.data()), extra_payload.size());
+	crypto_blake2b_final(&ctx, this->keys.data());
+	return true;
+}
+
+/**
+ * Encryption handler implementation for monocypther encryption after a X25519 key exchange.
+ */
+class X25519EncryptionHandler : public NetworkEncryptionHandler {
+private:
+	crypto_aead_ctx context; ///< The actual encryption context.
+
+public:
+	/**
+	 * Create the encryption handler.
+	 * @param key The key used for the encryption.
+	 * @param nonce The nonce used for the encryption.
+	 */
+	X25519EncryptionHandler(const std::span<const uint8_t> key, const X25519Nonce &nonce)
+	{
+		assert(key.size() == X25519_KEY_SIZE);
+		crypto_aead_init_x(&this->context, key.data(), nonce.data());
+	}
+
+	/** Ensure the encryption context is wiped! */
+	~X25519EncryptionHandler()
+	{
+		crypto_wipe(&this->context, sizeof(this->context));
+	}
+
+	size_t MACSize() const override
+	{
+		return X25519_MAC_SIZE;
+	}
+
+	bool Decrypt(std::span<std::uint8_t> mac, std::span<std::uint8_t> message) override
+	{
+		return crypto_aead_read(&this->context, message.data(), mac.data(), nullptr, 0, message.data(), message.size()) == 0;
+	}
+
+	void Encrypt(std::span<std::uint8_t> mac, std::span<std::uint8_t> message) override
+	{
+		crypto_aead_write(&this->context, message.data(), mac.data(), nullptr, 0, message.data(), message.size());
+	}
+};
+
+/** Ensure the key does not get leaked when we're done with it. */
+X25519Key::~X25519Key()
+{
+	crypto_wipe(*this);
+}
+
+/**
+ * Create a new secret key that's filled with random bytes.
+ * @return The randomly filled key.
+ */
+/* static */ X25519SecretKey X25519SecretKey::CreateRandom()
+{
+	X25519SecretKey secret_key;
+	RandomBytesWithFallback(secret_key);
```

Comment:
`WithFallback` doesn't look appropriate for cryptography.

Reply:
Technically true, but at some point there is a limit what you can do. That function will, under our supported operating systems return cryptographically-strong random.

The major difference between your (libhydrogen) and our implementation, for the supported platforms, is that your implementation `abort()`s/crashes without useful message to the user the whole application and ours falls back to a worse random when for some  reason the OS does not return random data.

Note that our implementation follows the manual of Monocypher for our supported platforms. So it could only be a problem on some esoteric platform, or under very weird circumstance.